### PR TITLE
More comments

### PR DESCRIPTION
**Pull request by Brookoko**

### CODE REVIEW

File: proto/scene.proto [hunk 1]
```diff
@@ -58,19 +85,52 @@ Can be overridden in renderer
...
   int32 width = 2;
+  // Height of final image
   int32 height = 3;
 }
 
-// Representation of cube 
-message Cube {}
+/*
+Representation of cube
+To determine the position of a sphere look at Transform.position
```
Comment (MetaMmodern) [inline proto/scene.proto]: `position of a sphere`
cube maybe?)

**Review by MetaMmodern — CHANGES_REQUESTED** [review]

**Review by MetaMmodern — APPROVED** [review]

File: proto/scene.proto [hunk 2]
```diff
@@ -22,23 +22,50 @@ message Vector3 {
...
 }
 
-// Position, rotation and scale of an object
+/*
+Position, rotation and scale of an object
+To calculate transformation matrix
+create translation matrix using position
+create rotation matrix using rotation
+create scale matrix using scale
+and multiple 3 matrices to get final matrix
```
Comment (nikitavbv) [inline proto/scene.proto]: This is way more clear now :+1:.
Maybe format this as a list? Add some bullet points or HTML markup?

File: proto/scene.proto [hunk 3]
```diff
@@ -22,23 +22,50 @@ message Vector3 {
...
+Process using parent is specified below
+
+Right vector can be extract
```
Comment (nikitavbv) [inline proto/scene.proto]: extracted

File: proto/scene.proto [hunk 4]
```diff
@@ -22,23 +22,50 @@ message Vector3 {
...
+from 1 column of the transformation matrix if your renderer use column-major conversion
```
Comment (nikitavbv) [inline proto/scene.proto]: 1st, 2nd, ... instead of 1, 2
renderer `uses`

File: proto/scene.proto [hunk 5]
```diff
@@ -22,23 +22,50 @@ message Vector3 {
...
+from 2 column of the transformation matrix if your renderer use column-major conversion 
+from 2 row of the transformation matrix if your renderer use row-major conversion
+
+Forward vector can be extract
+from 3 column of the transformation matrix if your renderer use column-major conversion 
+from 3 row of the transformation matrix if your renderer use row-major conversion 
+*/
 message Transform {
+  /*
+  If not specified default value will be (0, 0, 0)
```
Comment (nikitavbv) [inline proto/scene.proto]: What do you think about writing something like this:
```
@default (0, 0, 0)
````

And we will hack our HTML template to have another table column with a default value?

File: proto/scene.proto [hunk 6]
```diff
@@ -58,19 +85,52 @@ Can be overridden in renderer
 message RenderOptions {
   // Camera from which scene should be rendered
   int32 camera_id = 1;
+  // Width of final image
```
Comment (nikitavbv) [inline proto/scene.proto]: What if we specify defaults for width and height?

File: proto/scene.proto [hunk 7]
```diff
@@ -58,19 +85,52 @@ Can be overridden in renderer
...
 
-// Representation of cube 
-message Cube {}
+/*
+Representation of cube
+To determine the position of a cube look at Transform.position
+Transform.scale should be ignored
+If your renderer supports only axis aligned box Transform.rotation should be ignored
+For those who dare to support oriented bounding box here link how obb is implemented in opengl:
+http://www.opengl-tutorial.org/miscellaneous/clicking-on-objects/picking-with-custom-ray-obb-function/
```
Comment (nikitavbv) [inline proto/scene.proto]: Just a reminder that HTML formatting works, so you can use `<a href="...">` to make links clickable in the doc.

File: proto/scene.proto [hunk 8]
```diff
@@ -58,19 +85,52 @@ Can be overridden in renderer
...
-// Representation of sphere 
-message Sphere {}
+/*
+Representation of sphere
+Transform.scale should be ignored
+To determine the position of a sphere look at Transform.position
+*/
+message Sphere {
+  // Radius of sphere
+  float radius = 1;
```
Comment (nikitavbv) [inline proto/scene.proto]: We use doubles everywhere.

File: proto/scene.proto [hunk 9]
```diff
@@ -58,19 +85,52 @@ Can be overridden in renderer
...
+}
 
-// Representation of plane 
+/*
+Representation of plane
```
Comment (nikitavbv) [inline proto/scene.proto]: Based on our discussion in that chat, it seems that a plane should be defined. We should make it more clear what a plane is.

File: proto/scene.proto [hunk 10]
```diff
@@ -58,19 +85,52 @@ Can be overridden in renderer
...
+To determine the position of a plane look at Transform.position
+To determine the normal of a plane look at up vector of transformation
```
Comment (nikitavbv) [inline proto/scene.proto]: Vector of transformation = Transform.rotation?

File: proto/scene.proto [hunk 11]
```diff
@@ -92,28 +152,29 @@ message SceneObject {
...
-    // @oneof_item
-    Cube cube = 6;
-    // @oneof_item
-    Plane plane = 7;
-    // @oneof_item
-    MeshedObject meshed_object = 8;
+    Sphere sphere = 5; // @oneof 4 mesh
+    Cube cube = 6; // @oneof_item
+    Plane plane = 7; // @oneof_item
+    MeshedObject meshed_object = 8; // @oneof_item
```
Comment (nikitavbv) [inline proto/scene.proto]: You forgot to add `Disk` here.

File: proto/scene.proto (hunk 5 above)
Comment (Brookoko) [reply proto/scene.proto]: That sounds good. In this column, we also can mention which field is mandatory to specify

File: proto/scene.proto (hunk 10 above)
Comment (Brookoko) [reply proto/scene.proto]: As mentioned in the Transform section up vector can be extracted from matrix

File: proto/scene.proto [hunk 12]
```diff
@@ -92,38 +182,54 @@ message SceneObject {
...
   oneof mesh {
-    // @oneof 4 mesh
-    Sphere sphere = 5;
-    // @oneof_item
-    Cube cube = 6;
-    // @oneof_item
-    Plane plane = 7;
-    // @oneof_item
-    MeshedObject meshed_object = 8;
+    Sphere sphere = 5; // @oneof 4 mesh
```
Comment (nikitavbv) [inline proto/scene.proto]:
```
@oneof 5
```

File: proto/scene.proto [hunk 13]
```diff
@@ -134,25 +240,41 @@ message Light {
...
+  Color of source light
+  */
   Color color = 4;
-  double intensity = 5;
   oneof light {
-    LightPoint point = 6; // @oneof 4 light
-    LightDirectional directional = 7; // @oneof_item
-    LightSphere sphere = 8; // @oneof_item @optional
-    LightSurrounding surrounding = 9; // @oneof_item
+    LightPoint point = 5; // @oneof 4 light
```
Comment (nikitavbv) [inline proto/scene.proto:249]: It is okay for now, but in the future, we should never change the fields number.

File: proto/scene.proto [hunk 14]
```diff
@@ -134,25 +240,41 @@ message Light {
...
   }
 }
 
 // Representation of perspective camera
-message PerspectiveCamera { double fov = 1; }
+message PerspectiveCamera {
+  /*
+  @default 60
+  Horizontal field of view
+  Should be in range [0, 179]
```
Comment (nikitavbv) [inline proto/scene.proto]: (0, 180)
?

File: proto/scene.proto [hunk 15]
```diff
@@ -162,27 +284,45 @@ message Camera {
...
+or this <a href="https://pbr-book.org/3ed-2018/Reflection_Models">link</a>
+*/
 message Material {
   string id = 1;
   oneof material {
-    // @oneof 2 material
-    SolidMaterial solid = 2;
-    // @oneof_item
-    MirrorMaterial glass = 3;
+    LambertReflectionMaterial lambertReflection = 2; // @oneof 2 material
```
Comment (nikitavbv) [inline proto/scene.proto]: Please use snake_case. Protobuf will translate it to the right style for each language (i.e. camelCase for C#)

**Review by nikitavbv — APPROVED** [review]

**Pull request merged**